crates/pink-libs/sidevmop/lib.rs

```diff
@@ -0,0 +1,86 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+
+use pink_extension as pink;
+
+#[pink::contract(env = PinkEnvironment)]
+mod sidevmop {
+    use super::pink;
+    use ink_storage::{traits::SpreadAllocate, Mapping};
+    use pink::system::{Error, Result};
+    use pink::PinkEnvironment;
+
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub struct SidevmOp {
+        /// Owner of the contract
+        owner: AccountId,
+        /// Contracts that are allowed to deploy sidevm.
+        whitelist: Mapping<AccountId, ()>,
+    }
+
+    impl SidevmOp {
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            ink_lang::utils::initialize_contract(|me: &mut Self| {
+                me.owner = Self::env().caller();
+            })
+        }
+        #[ink(message)]
+        pub fn allow(&mut self, contract: AccountId) -> Result<()> {
+            if self.env().caller() != self.owner {
+                return Err(Error::BadOrigin);
+            }
+            self.whitelist.insert(contract, &());
+            Ok(())
+        }
+    }
+
+    impl pink::system::SidevmOperation for SidevmOp {
+        #[ink(message)]
+        fn deploy(&self, code_hash: pink::Hash) -> Result<()> {
+            let caller = self.env().caller();
+            if !self.whitelist.contains(&caller) {
+                return Err(Error::BadOrigin);
+            }
+            let system = pink::system::SystemRef::instance();
+            system.deploy_sidevm_to(caller, code_hash)
```

Comment:
To get `System` to allow the `deploy_sidevm_to` call, it's necessary to add the driver (`SidevmOp`) as an admin, right? I didn't find where it's called in unit tests.

Reply:
> it's necessary to add the driver (SidevmOp) as an admin, right? 

Yeah, or the owner. Should we forbid the owner to invoke deploy_sidevm_to?

> I didn't find where it's called in unit tests.

The caller is always [1u8; 32] in the unit test, which is already the owner.

Reply:
Refactored the unittests